src/setfit/onnx.py

```diff
@@ -0,0 +1,347 @@
+    :param opset: The ONNX opset to use for optimizing this model. The opset is not
+        gauranteed and will default to the maximum version possible for the sklearn
+        model.
+    :raises ImportError: If skl2onnx is not installed an error will be raised asking
+        to install this package.
+    :return: The ONNX model generated from the sklearn head.
+    """
+
+    # Check if skl2onnx is installed
+    try:
+        from skl2onnx.common.data_types import guess_data_type
+        from skl2onnx import convert_sklearn
+        import onnxconverter_common
+        from skl2onnx.sklapi import CastTransformer
+        from sklearn.pipeline import Pipeline
+    except ImportError as e:
+        msg = "skl2onnx must be installed in order to convert a model with an sklearn head."
+        raise ImportError(msg)
+
+    # Determine the initial type and the shape of the output.
+    input_shape = (None, *model.model_head.coef_.shape[1:])
+    dtype = guess_data_type(model.model_head.coef_, shape=input_shape)[0][1]
+    dtype.shape = input_shape
+
+    # If the datatype of the model is double we need to cast the outputs
+    # from the setfit model to doubles for compatibility inside of ONNX.
+    if type(dtype) == onnxconverter_common.data_types.DoubleTensorType:
+        # TODO:: TALK ABOUT FLOAT CONVERSION ISSUES.
+        sk_model = Pipeline(
+            [
+                ("castdouble", CastTransformer(dtype=np.double)),
+                ("head", model.model_head),
+            ]
+        )
+    else:
+        sk_model = model.model_head
+
+    # Convert sklearn head into ONNX format
+    initial_type = [("model_head", dtype)]
+    onx = convert_sklearn(
+        sk_model,
+        initial_types=initial_type,
+        target_opset=opset,
+        options={id(sk_model): {"zipmap": False}},
+    )
+
+    return onx
+
+
+def export_onnx(
+    model_name: str, opset: int, output: Path, use_external_format: bool = False
+):
+    """
+    Export a PyTorch backed setfit model to ONNX Intermediate Representation.
+
+    Args:
+        model: The name of the pretrained setfit model to load and use. It can be a path
+            to a locally trained model.
+        opset: The actual version of the ONNX operator set to use.  The final opset used
+            might be lower. ONNX will use the highest version supported by both the
+            sklearn head and the model body. If versions can't be rectified an error
+            will be thrown.
+        output: Path where will be stored the generated ONNX model. If the head is an
+            sklearn.estimator the models will be saved in
+        use_external_format: Split the model definition from its parameters to allow
+            model bigger than 2GB
+    Returns:
+    """
+
+    model = SetFitModel.from_pretrained(model_name)
+
+    # Check to see if the model uses a sklearn head or a torch dense layer.
+    if issubclass(type(model.model_head), models.Dense):
```

Comment:
How about before this if-else statement, we extract `model` and `pooler` as mentioned in this [notebook](https://drive.google.com/file/d/1KvN95i6vZlBW-eecq_waQiCwjO3vPvFR/view?usp=share_link)? So you can continue on connecting those two with the `sklearn` head and I can do so on the `pytorch` head @nbertagnolli

Reply:
This is awesome! Yeah pulling the pooler out and attaching them like this is way better.   I'll update the PR. Thanks! 
🙏 @blakechi 🙏

Reply:
Glad you like it!
ok, I will add my park as well after your update

Reply:
I added in a bunch of your suggestions : ). Thanks for all the great input! Let me know if there is anything else you see that could be better.

Reply:
Added my part at here: nbertagnolli/setfit#1